execution/group_by.go

```diff
@@ -176,20 +210,47 @@ func (gb *GroupByStream) Trigger(ctx context.Context, tx storage.StateTransactio
 		return nil, errors.Wrap(err, "couldn't get previously triggered value for key")
 	}
 
-	values := make([]octosql.Value, len(gb.aggregates))
-	for i := range gb.aggregates {
-		var err error
-		values[i], err = gb.aggregates[i].GetValue(ctx, txByKey.WithPrefix(gb.prefixes[i]))
+	// Check if record count == retraction count
+	recordCountState := storage.NewValueState(txByKey.WithPrefix(recordCountPrefix))
+	var recordCountValue octosql.Value
+	err = recordCountState.Get(&recordCountValue)
+	if err == storage.ErrNotFound {
+		recordCountValue = octosql.MakeInt(0)
+	} else if err != nil {
+		return nil, errors.Wrap(err, "couldn't get current record count")
+	}
+
+	if recordCountValue.AsInt() > 0 {
+		// Get new record to trigger
+		values := make([]octosql.Value, len(gb.aggregates))
+		for i := range gb.aggregates {
+			var err error
+			values[i], err = gb.aggregates[i].GetValue(ctx, txByKey.WithPrefix(gb.prefixes[i]))
+			if err != nil {
+				return nil, errors.Wrapf(err, "couldn't get value of aggregate %s with index %d", gb.aggregates[i].String(), i)
+			}
+		}
+		output = append(output, NewRecordFromSlice(gb.outputFieldNames, values, opts...))
+
+		// Save currently triggered record
+		newPreviouslyTriggeredValuesTuple := octosql.MakeTuple(values)
+		err = previouslyTriggeredValues.Set(&newPreviouslyTriggeredValuesTuple)
+		if err != nil {
+			return nil, errors.Wrap(err, "couldn't set new previously triggered value")
+		}
+	} else {
+		// Clear previously triggered record, as we're sending a retraction for it now
+		err = previouslyTriggeredValues.Clear()
 		if err != nil {
-			return nil, errors.Wrapf(err, "couldn't get value of aggregate %s with index %d", gb.aggregates[i].String(), i)
+			return nil, errors.Wrap(err, "couldn't clear previously triggered value")
 		}
 	}
-	output = append(output, NewRecordFromSlice(gb.outputFieldNames, values, opts...))
 
-	newPreviouslyTriggeredValuesTuple := octosql.MakeTuple(values)
-	err = previouslyTriggeredValues.Set(&newPreviouslyTriggeredValuesTuple)
-	if err != nil {
-		return nil, errors.Wrap(err, "couldn't set new previously triggered value")
+	if len(output) == 2 {
```

Comment:
You could add a comment here, not very clear for me what is going on :<

Reply:
added